File: trunk/incubators/net.sf.j2s.java.org.eclipse.osgi/src/org/eclipse/core/runtime/internal/adaptor/BundleStopper.java
/*******************************************************************************
 * Copyright (c) 2003, 2004 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 * 
 * Contributors:
 *     IBM Corporation - initial API and implementation
 *******************************************************************************/
package org.eclipse.core.runtime.internal.adaptor;

import java.util.Hashtable;
import org.eclipse.core.runtime.adaptor.*;
import org.eclipse.osgi.framework.adaptor.FrameworkAdaptor;
import org.eclipse.osgi.framework.debug.Debug;
import org.eclipse.osgi.framework.internal.core.AbstractBundle;
import org.eclipse.osgi.framework.internal.core.BundleHost;
import org.eclipse.osgi.framework.log.FrameworkLogEntry;
import org.eclipse.osgi.service.resolver.BundleDescription;
import org.eclipse.osgi.service.resolver.StateHelper;
import org.eclipse.osgi.util.NLS;
import org.osgi.framework.Bundle;
import org.osgi.framework.BundleContext;

/**
 * Implementation for the runtime shutdown hook that provides 
 * support for legacy bundles. All legacy bundles are stopped 
 * in the proper order.
 * 
 * <p>Internal class.</p>
 */
public class BundleStopper {
	/* must be a synchronized object */
	private Hashtable stoppedBundles;

	private BundleDescription[] allToStop = null;

	private BundleContext context;

	public BundleStopper(BundleContext context) {
		this.context = context;
	}

	private void logCycles(Object[][] cycles) {
		if (! Debug.DEBUG)
			return;
		
		// log cycles
		if (cycles.length > 0) {
			StringBuffer cycleText = new StringBuffer("["); //$NON-NLS-1$			
			for (int i = 0; i < cycles.length; i++) {
				cycleText.append('[');
				for (int j = 0; j < cycles[i].length; j++) {
					cycleText.append(((BundleDescription) cycles[i][j]).getSymbolicName());
					cycleText.append(',');
				}
				cycleText.insert(cycleText.length() - 1, ']');
			}
			cycleText.setCharAt(cycleText.length() - 1, ']');
			String message = NLS.bind(EclipseAdaptorMsg.ECLIPSE_BUNDLESTOPPER_CYCLES_FOUND, cycleText); //$NON-NLS-1$
			FrameworkLogEntry entry = new FrameworkLogEntry(FrameworkAdaptor.FRAMEWORK_SYMBOLICNAME, message, 0, null, null);
			EclipseAdaptor.getDefault().getFrameworkLog().log(entry);
		}
	}

	public void stopBundles() {
		allToStop = EclipseAdaptor.getDefault().getState().getResolvedBundles();
		StateHelper stateHelper = EclipseAdaptor.getDefault().getPlatformAdmin().getStateHelper();
		Object[][] cycles = stateHelper.sortBundles(allToStop);
		logCycles(cycles);
		stoppedBundles = new Hashtable(allToStop.length);
		basicStopBundles();
	}

	private void basicStopBundles() {
		// stop all active bundles in the reverse order of Require-Bundle
		for (int stoppingIndex = allToStop.length - 1; stoppingIndex >= 0; stoppingIndex--) {
			AbstractBundle toStop = (AbstractBundle) context.getBundle(allToStop[stoppingIndex].getBundleId());
			if (toStop.getBundleId() != 0 && ((EclipseBundleData) toStop.getBundleData()).isAutoStartable()) {
				try {
					if ((toStop.getState() == Bundle.ACTIVE) && (toStop instanceof BundleHost))
						toStop.stop();
				} catch (Exception e) {
					String message = NLS.bind(EclipseAdaptorMsg.ECLIPSE_BUNDLESTOPPER_ERROR_STOPPING_BUNDLE, allToStop[stoppingIndex].toString());
					FrameworkLogEntry entry = new FrameworkLogEntry(FrameworkAdaptor.FRAMEWORK_SYMBOLICNAME, message, 0, e, null);
					EclipseAdaptor.getDefault().getFrameworkLog().log(entry);
				} finally {
					stoppedBundles.put(toStop, toStop);
				}
			}
		}
	}

	public boolean isStopped(Bundle bundle) {
		if (stoppedBundles == null)
			return false;
		return stoppedBundles.get(bundle) != null;
	}
}

File: trunk/incubator/net.sf.j2s.test.mauve/j2slib/org/eclipse/core/runtime/internal/adaptor/BundleStopper.js
Clazz.declarePackage ("org.eclipse.core.runtime.internal.adaptor");
Clazz.load (null, "org.eclipse.core.runtime.internal.adaptor.BundleStopper", ["java.lang.StringBuffer", "java.util.Hashtable", "org.eclipse.core.runtime.adaptor.EclipseAdaptor", "$.EclipseAdaptorMsg", "org.eclipse.osgi.framework.log.FrameworkLogEntry", "org.eclipse.osgi.util.NLS"], function () {
c$ = Clazz.decorateAsClass (function () {
this.stoppedBundles = null;
this.allToStop = null;
this.context = null;
Clazz.instantialize (this, arguments);
}, org.eclipse.core.runtime.internal.adaptor, "BundleStopper");
Clazz.makeConstructor (c$, 
function (context) {
this.context = context;
}, "org.osgi.framework.BundleContext");
Clazz.defineMethod (c$, "logCycles", 
($fz = function (cycles) {
if (false) return ;
if (cycles.length > 0) {
var cycleText =  new StringBuffer ("[");
for (var i = 0; i < cycles.length; i++) {
cycleText.append ('[');
for (var j = 0; j < cycles[i].length; j++) {
cycleText.append ((cycles[i][j]).getSymbolicName ());
cycleText.append (',');
}
cycleText.insert (cycleText.length () - 1, ']');
}
cycleText.setCharAt (cycleText.length () - 1, ']');
var message = org.eclipse.osgi.util.NLS.bind (org.eclipse.core.runtime.adaptor.EclipseAdaptorMsg.ECLIPSE_BUNDLESTOPPER_CYCLES_FOUND, cycleText);
var entry =  new org.eclipse.osgi.framework.log.FrameworkLogEntry ("org.eclipse.osgi", message, 0, null, null);
org.eclipse.core.runtime.adaptor.EclipseAdaptor.getDefault ().getFrameworkLog ().log (entry);
}}, $fz.isPrivate = true, $fz), "~A");
Clazz.defineMethod (c$, "stopBundles", 
function () {
this.allToStop = org.eclipse.core.runtime.adaptor.EclipseAdaptor.getDefault ().getState ().getResolvedBundles ();
var stateHelper = org.eclipse.core.runtime.adaptor.EclipseAdaptor.getDefault ().getPlatformAdmin ().getStateHelper ();
var cycles = stateHelper.sortBundles (this.allToStop);
this.logCycles (cycles);
this.stoppedBundles =  new java.util.Hashtable (this.allToStop.length);
this.basicStopBundles ();
});
Clazz.defineMethod (c$, "basicStopBundles", 
($fz = function () {
for (var stoppingIndex = this.allToStop.length - 1; stoppingIndex >= 0; stoppingIndex--) {
var toStop = this.context.getBundle (this.allToStop[stoppingIndex].getBundleId ());
if (toStop.getBundleId () != 0 && (toStop.getBundleData ()).isAutoStartable ()) {
try {
if ((toStop.getState () == 32) && (Clazz.instanceOf (toStop, org.eclipse.osgi.framework.internal.core.BundleHost))) toStop.stop ();
} catch (e) {
if (Clazz.instanceOf (e, Exception)) {
var message = org.eclipse.osgi.util.NLS.bind (org.eclipse.core.runtime.adaptor.EclipseAdaptorMsg.ECLIPSE_BUNDLESTOPPER_ERROR_STOPPING_BUNDLE, this.allToStop[stoppingIndex].toString ());
var entry =  new org.eclipse.osgi.framework.log.FrameworkLogEntry ("org.eclipse.osgi", message, 0, e, null);
org.eclipse.core.runtime.adaptor.EclipseAdaptor.getDefault ().getFrameworkLog ().log (entry);
} else {
throw e;
}
} finally {
this.stoppedBundles.put (toStop, toStop);
}
}}
}, $fz.isPrivate = true, $fz));
Clazz.defineMethod (c$, "isStopped", 
function (bundle) {
if (this.stoppedBundles == null) return false;
return this.stoppedBundles.get (bundle) != null;
}, "org.osgi.framework.Bundle");
});
